File: src/main/kotlin/dev/jefy/connectpro/portfolio/applicaion/command/ProjectCommandImpl.java
package dev.jefy.connectpro.portfolio.applicaion.command;


import org.jspecify.annotations.NullMarked;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.Assert;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;

import dev.jefy.connectpro.portfolio.applicaion.dtos.ProjectRequest;
import dev.jefy.connectpro.portfolio.domain.model.Project;
import dev.jefy.connectpro.portfolio.domain.repository.PortfolioRepository;
import dev.jefy.connectpro.portfolio.domain.repository.ProjectRepository;
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId;
import dev.jefy.connectpro.portfolio.domain.vo.ProjectId;
import dev.jefy.connectpro.shared.application.exceptions.ResourceNotFound;
import dev.jefy.connectpro.shared.domain.vo.ImageUrl;
import dev.jefy.connectpro.shared.infrastructure.file_storage.ImageService;
import lombok.RequiredArgsConstructor;

/**
 * @author Jôph Yamba
 */
@NullMarked
@Service
@Transactional
@RequiredArgsConstructor
public class ProjectCommandImpl implements ProjectCommand {
    private final PortfolioRepository portfolioRepo;
    private final ProjectRepository projectRepo;
    private final ImageService imageService;

    @Override
    public ProjectId create(ProjectRequest request) {
        PortfolioId portfolioId = PortfolioId.of(request.portfolioId());
        if (!portfolioRepo.existsById(portfolioId)) {
            throw  new ResourceNotFound("Portfolio with id " + portfolioId + " not found");
        }
        boolean isConflict = projectRepo.isTitleConflict(portfolioId, request.title());
        if (isConflict) {
            throw new IllegalStateException("Project with title already exists");
        }
        Project project = new Project(portfolioId, request);
        projectRepo.save(project);
        return project.getId();
    }

    @Override
    public ProjectId update(ProjectId projectId, ProjectRequest request) {
        Project project = getProject(projectId);
        boolean isConflict = projectRepo.isTitleConflict(project.getPortfolioId(), request.title());
        if (isConflict) {
            throw new IllegalStateException("project with title already exists");
        }
        project.update(request);
        projectRepo.save(project);
        return project.getId();
    }

    @Override
    public ProjectId addImage(ProjectId projectId, MultipartFile image) throws IOException {
        Assert.notNull(image, "Image must not be null");
        Project project = getProject(projectId);

        ImageUrl imageUrl = imageService.save(image);
        project.addImage(imageUrl);
        projectRepo.save(project);
        return project.getId();
    }

    @Override
    public ProjectId removeImage(ProjectId projectId, ImageUrl imageUrl) throws IOException {
        Assert.notNull(imageUrl, "Image must not be null");
        Project project = getProject(projectId);
        imageService.delete(imageUrl);
        project.removeImage(imageUrl);
        projectRepo.save(project);
        return project.getId();
    }

    private Project getProject(ProjectId projectId) {
        Assert.notNull(projectId, "Project id cannot be null");
        return projectRepo
                .findById(projectId)
                .orElseThrow(() -> new ResourceNotFound("Project not found"));
    }
}

File: src/main/kotlin/dev/jefy/connectpro/portfolio/application/command/ProjectCommandImpl.kt
package dev.jefy.connectpro.portfolio.application.command

import dev.jefy.connectpro.portfolio.application.dtos.ProjectRequest
import dev.jefy.connectpro.portfolio.application.exceptions.ProjectAlreadyExistsException
import dev.jefy.connectpro.portfolio.application.exceptions.ProjectNotFoundException
import dev.jefy.connectpro.portfolio.domain.model.Project
import dev.jefy.connectpro.portfolio.domain.repository.PortfolioRepository
import dev.jefy.connectpro.portfolio.domain.repository.ProjectRepository
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId
import dev.jefy.connectpro.portfolio.domain.vo.ProjectId
import dev.jefy.connectpro.shared.application.dtos.ImageData
import dev.jefy.connectpro.shared.domain.vo.Image
import dev.jefy.connectpro.shared.infrastructure.annotations.CommandService
import dev.jefy.connectpro.shared.infrastructure.file_storage.ImageService
import org.springframework.web.multipart.MultipartFile
import java.io.IOException

@CommandService
class ProjectCommandImpl(
    private val portfolioRepo: PortfolioRepository,
    private val projectRepo: ProjectRepository,
    private val imageService: ImageService
) : ProjectCommand {

    override fun create(request: ProjectRequest): ProjectId {
        val portfolioId = PortfolioId.of(request.portfolioId)

        check(portfolioRepo.existsById(portfolioId)) { "Portfolio with id $portfolioId not found" }

        val isConflict = projectRepo.isTitleConflict(
            portfolioId = portfolioId,
            title = request.name,
        )
        if (isConflict) throw ProjectAlreadyExistsException()

        val project = Project(
            portfolioId = portfolioId,
            name = request.name,
            description = request.description,
        )
        projectRepo.save(project)

        return project.id
    }

    override fun update(projectId: ProjectId, request: ProjectRequest): ProjectId {
        val project = getProject(projectId)
        val isConflict = projectRepo.isTitleConflictForId(
            portfolioId = project.portfolioId,
            title = request.name,
            projectId = projectId
        )
        if (isConflict) throw ProjectAlreadyExistsException()
        project.update(request)
        projectRepo.save(project) 
        return project.id
    }

    @Throws(IOException::class)
    override fun addImage(projectId: ProjectId, image: MultipartFile): ProjectId {
        val project = getProject(projectId)
        val imageUrl = imageService.save(image)
        project.addImage(imageUrl)
        projectRepo.save(project)
        return project.id
    }
    
    @Throws(IOException::class)
    override fun addImages(projectId: ProjectId, images: List<MultipartFile>): ProjectId {
        val project = getProject(projectId)
        images.forEach { file -> 
            val imageUrl = imageService.save(file)
            project.addImage(imageUrl)
        }
        projectRepo.save(project)
        return project.id
    }

    @Throws(IOException::class)
    override fun removeImage(projectId: ProjectId, image: ImageData): ProjectId {
        val project = getProject(projectId)
        imageService.delete(image.getKey())
        project.removeImage(image.getKey())
        projectRepo.save(project)
        return project.id
    }
        
            

    @Throws(IOException::class)
    override fun delete(projectId: ProjectId) {
        val project = getProject(projectId)
        project.images.forEach { imageService.delete(Image(it)) }
        projectRepo.delete(project)
    }

    private fun getProject(projectId: ProjectId): Project = projectRepo
        .findById(projectId).orElseThrow { ProjectNotFoundException() }
    
}